### PR TITLE
Stringifier expansion

### PR DESCRIPTION
**Pull request by IlliaShepilov**
Summary of changes:
- `Stringifier` is abstract class from now.
- `Stringifier` extends `com.google.common.base.Converter` and can make reversal conversion from now.
- If `null` is passed to the `convert` method, `null` will be returned.

### CODE REVIEW

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/base/Stringifier.java [hunk 1]
```diff
@@ -20,12 +20,13 @@
...
 
 /**
- * An object converting to {@code String}.
+ * An object converting from A to B and reverse.
  *
  * @author Alexander Yevsyukov
  */
-public interface Stringifier<T> extends Function<T, String> {
+@SuppressWarnings("AbstractMethodOverridesAbstractMethod")
+public abstract class Stringifier<A, B> extends Converter<A, B> {
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifier.java]: Describe the generic parameters in the Javadoc section.

File: client/src/main/java/org/spine3/base/Stringifier.java [hunk 2]
```diff
@@ -20,12 +20,13 @@
...
+import com.google.common.base.Converter;
 
 /**
- * An object converting to {@code String}.
+ * An object converting from A to B and reverse.
  *
  * @author Alexander Yevsyukov
  */
-public interface Stringifier<T> extends Function<T, String> {
+@SuppressWarnings("AbstractMethodOverridesAbstractMethod")
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifier.java]: Please review if you need this suppression.

File: client/src/main/java/org/spine3/base/Stringifier.java (hunk 1 above)
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifier.java]: B is a `String`, isn't it?

File: client/src/main/java/org/spine3/base/Stringifiers.java [hunk 3]
```diff
@@ -139,38 +146,63 @@ private static String convert(Message message) {
...
+        @SuppressWarnings("ThrowInsideCatchBlockWhichIgnoresCaughtException")
+        // It is OK, because all necessary information about exception
+        // is passed to the {@code ConversionError} instance.
+        protected Timestamp doBackward(String s) {
+            try {
+                return Timestamps.parse(s);
+            } catch (ParseException e) {
+                final ConversionError conversionError = new ConversionError(e.getMessage(),
+                                                                            e.getErrorOffset());
+                throw new IllegalArgumentException(conversionError);
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: That's weird. How will you fetch the conversion error from the outside?

File: client/src/main/java/org/spine3/validate/ConversionError.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.validate;
+
+import java.text.ParseException;
+
+/**
+ * Signals that an error has been reached unexpectedly while conversion from one type to another.
```
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConversionError.java]: `while converting`

File: client/src/main/java/org/spine3/validate/ConversionError.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @author Illia Shepilov
+ */
+@SuppressWarnings("ExceptionClassNameDoesntEndWithException")
+// It is OK, because it is not {@code Exception} in usual meaning.
+// It is occurred when input value cannot be converted to desirable type
+// and repeated input is required.
+public class ConversionError extends ParseException {
```
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConversionError.java]: Is a `ConversionError` a `ParseException` in fact? Please check the docs on the `ParseException`.

To me, the conversion error is a more general case, than what's described in the `ParseException` docs.

File: client/src/main/java/org/spine3/validate/ConversionError.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Constructs a ParseException with the specified detail message and
```
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConversionError.java]: Please format this section properly.

**Comment by armiol** [comment]
@IlliaShepilov 
```
If pass null to the convert method, null will be returned.
```
should be rephrased. Use passive voice for the first part as well.

File: client/src/main/java/org/spine3/base/Stringifier.java (hunk 2 above)
Comment (IlliaShepilov) [inline client/src/main/java/org/spine3/base/Stringifier.java]: Forgot to delete.

File: client/src/main/java/org/spine3/base/Stringifiers.java (hunk 3 above)
Comment (IlliaShepilov) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: With help of the `getCause()` method.

File: client/src/main/java/org/spine3/validate/ConversionError.java (hunk 5 above)
Comment (IlliaShepilov) [inline client/src/main/java/org/spine3/validate/ConversionError.java]: `Signals that an error has been reached unexpectedly while parsing.` - that is description of the `ParseException` class. It fits for our purposes, as for me.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/base/Stringifier.java [hunk 7]
```diff
@@ -20,12 +20,14 @@
 
 package org.spine3.base;
 
-import com.google.common.base.Function;
+import com.google.common.base.Converter;
 
 /**
- * An object converting to {@code String}.
+ * Converts an object of the I class to the {@code String} and reverse.
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifier.java]: `{@code I}`?

Also, let's reprhase `reverse` as it does not fit the sentence. Check the Guava's documentation on the same matter.

File: client/src/main/java/org/spine3/base/Stringifiers.java (hunk 3 above)
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: How would you know that a `ConversionError`?

Comment (IlliaShepilov) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: I understand your objections.
But I can't throw `ConversionError` explicitly, as it is checked exception and going on inside the overridden `doBackward` Guava method.  So the possible decision is catch all `RuntimeException`, check it with `instanceof` and rethrow.

File: client/src/main/java/org/spine3/validate/ConversionError.java (hunk 5 above)
Comment (IlliaShepilov) [inline client/src/main/java/org/spine3/validate/ConversionError.java]: I have looked all checked exceptions and have found some exceptions with appropriate name, but they have _strange_ packages and does not have documentation. I can create own hierarchy with one or few elements, but I am not sure that we need it.

**Comment by IlliaShepilov** [comment]
@armiol Please see my comments.

File: client/src/main/java/org/spine3/base/Stringifier.java [hunk 8]
```diff
@@ -20,12 +20,18 @@
...
 
 /**
- * An object converting to {@code String}.
+ * Serves as converter from {@code I} to {@code String} with an associated
+ * reverse function from {@code String} to {@code I}.
+ *
+ * <p>It is used for converting back and forth between the different
+ * representations of the same information.
  *
  * @author Alexander Yevsyukov
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifier.java]: Please append yourself as an author.

File: client/src/main/java/org/spine3/base/Stringifiers.java [hunk 9]
```diff
@@ -139,38 +147,62 @@ private static String convert(Message message) {
...
+            } catch (ParseException e) {
+                final ConversionError conversionError = new ConversionError(e.getMessage());
+                throw new IllegalConversionArgumentException(conversionError);
             }
-            return eventId.getUuid();
         }
     }
 
-    static class CommandIdStringifier implements Stringifier<CommandId> {
+    static class EventIdStringifier extends Stringifier<EventId> {
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: Please explain additionally why this class is exposed to `package-private`, and not marked as `protected` as the `TimestampIdStringifier` above. 

Same applies to the static class below.

File: client/src/main/java/org/spine3/validate/IllegalConversionArgumentException.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.validate;
+
+/**
+ * Thrown to indicate that a method has been passed an illegal or
+ * inappropriate argument during conversion from one type to another.
```
Comment (armiol) [inline client/src/main/java/org/spine3/validate/IllegalConversionArgumentException.java]: `...during the conversion...`

File: client/src/test/java/org/spine3/base/StringifiersShould.java [hunk 11]
```diff
@@ -189,57 +197,113 @@ public void convert_to_string_message_id_wrapped_in_Any() {
...
 
     @Test
+    public void convert_string_to_event_id() {
+        final String id = newUuid();
+        final EventId expected = EventId.newBuilder()
+                                        .setUuid(id)
+                                        .build();
+        final EventId actual = new Stringifiers.EventIdStringifier().reverse()
+                                                                    .convert(id);
+
```
Comment (armiol) [inline client/src/test/java/org/spine3/base/StringifiersShould.java]: Please remove an empty line.

**Comment by IlliaShepilov** [comment]
@armiol PTAL.

File: client/src/main/java/org/spine3/base/Stringifiers.java [hunk 12]
```diff
@@ -25,8 +25,11 @@
 import com.google.protobuf.MessageOrBuilder;
 import com.google.protobuf.Timestamp;
 import com.google.protobuf.util.Timestamps;
+import org.spine3.validate.ConversionError;
+import org.spine3.validate.IllegalConversionArgumentException;
 
 import javax.annotation.Nullable;
+import java.text.ParseException;
 import java.util.Collection;
 import java.util.regex.Pattern;
```
Comment (armiol) [inline client/src/main/java/org/spine3/base/Stringifiers.java]: Please append your name as another `@author` for this class as well.

**Pull request merged**